compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java

```diff
@@ -2891,9 +2911,17 @@ public void visit(BLangService serviceNode) {
                 dlog.error(attachExpr.pos, DiagnosticErrorCode.INVALID_LISTENER_ATTACHMENT);
             }
 
+            // Validate listener attachment based on attach-point of the service decl and second param of listener.
             if (exprType.getKind() == TypeKind.OBJECT) {
                 BObjectType listenerType = (BObjectType) exprType;
                 validateServicePathOnListener(serviceNode, attachExpr, listenerType);
+            } else if (exprType.getKind() == TypeKind.UNION) {
+                for (BType memberType : ((BUnionType) exprType).getMemberTypes()) {
+                    if (memberType.tag == TypeTags.ERROR) {
+                        continue;
+                    }
+                    validateServicePathOnListener(serviceNode, attachExpr, (BObjectType) memberType);
```

Comment:
The following results in a bad, sad error here.

```ballerina
public class Listener {

    public isolated function 'start() returns error? { }

    public isolated function gracefulStop() returns error? { }

    public isolated function immediateStop() returns error? { }

    public isolated function detach(service object { isolated function foo(); } s) returns error? { }

    public isolated function attach(service object { isolated function foo(); } s, string[]|string? name = ()) returns error? { }

    isolated function register(service object {} s, string[]|string? name) returns error? { }
}

type ListenerType int|Listener;

listener ListenerType ln = new Listener();

service / on ln {

}
```

```log
[2021-01-14 00:26:20,469] SEVERE {b7a.log.crash} - class org.wso2.ballerinalang.compiler.semantics.model.types.BType cannot be cast to class org.wso2.ballerinalang.compiler.semantics.model.types.BObjectType (org.wso2.ballerinalang.compiler.semantics.model.types.BType and org.wso2.ballerinalang.compiler.semantics.model.types.BObjectType are in unnamed module of loader 'app') 
java.lang.ClassCastException: class org.wso2.ballerinalang.compiler.semantics.model.types.BType cannot be cast to class org.wso2.ballerinalang.compiler.semantics.model.types.BObjectType (org.wso2.ballerinalang.compiler.semantics.model.types.BType and org.wso2.ballerinalang.compiler.semantics.model.types.BObjectType are in unnamed module of loader 'app')
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:3019)
	at org.wso2.ballerinalang.compiler.tree.BLangService.accept(BLangService.java:142)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:3250)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:3218)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeDef(SemanticAnalyzer.java:3210)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:291)
	at org.wso2.ballerinalang.compiler.tree.BLangPackage.accept(BLangPackage.java:163)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyze(SemanticAnalyzer.java:256)
	at io.ballerina.projects.internal.CompilerPhaseRunner.typeCheck(CompilerPhaseRunner.java:217)
	at io.ballerina.projects.internal.CompilerPhaseRunner.performTypeCheckPhases(CompilerPhaseRunner.java:115)
	at io.ballerina.projects.ModuleContext.compileInternal(ModuleContext.java:348)
	at io.ballerina.projects.ModuleCompilationState$1.compile(ModuleCompilationState.java:45)
	at io.ballerina.projects.ModuleCompilationState$1.generatePlatformSpecificCode(ModuleCompilationState.java:53)
	at io.ballerina.projects.ModuleContext.generatePlatformSpecificCode(ModuleContext.java:301)
	at io.ballerina.projects.JBallerinaBackend.performCodeGen(JBallerinaBackend.java:131)
	at io.ballerina.projects.JBallerinaBackend.<init>(JBallerinaBackend.java:117)
	at io.ballerina.projects.JBallerinaBackend.lambda$from$0(JBallerinaBackend.java:93)
	at java.base/java.util.HashMap.computeIfAbsent(HashMap.java:1133)
	at io.ballerina.projects.PackageCompilation.getCompilerBackend(PackageCompilation.java:126)
	at io.ballerina.projects.JBallerinaBackend.from(JBallerinaBackend.java:92)
	at io.ballerina.cli.task.CompileTask.execute(CompileTask.java:69)
	at io.ballerina.cli.TaskExecutor.executeTasks(TaskExecutor.java:40)
	at io.ballerina.cli.cmd.RunCommand.execute(RunCommand.java:154)
	at java.base/java.util.Optional.ifPresent(Optional.java:183)
	at io.ballerina.cli.launcher.Main.main(Main.java:58)
```

Reply:
Fixed. Thank you.